lib/ecto/changeset.ex

```diff
@@ -2970,8 +2970,8 @@ defmodule Ecto.Changeset do
 
     acc = Map.put(acc, key, relation_changed)
 
-    case relation do
-      %Ecto.Association.BelongsTo{} -> Map.put(acc, relation.owner_key, relation_changed.id)
+    case {relation, relation_changed} do
+      {%Ecto.Association.BelongsTo{}, %{id: id}} -> Map.put(acc, relation.owner_key, id)
```

Comment:
Unfortunately we cannot assume the primary key is id. However, in this case, your point is that `relation_changed` is nil, right? I think checking only for a map should be enough then?

```suggestion
      {%Ecto.Association.BelongsTo{}, %{}} -> Map.put(acc, relation.owner_key, id)
```

Reply:
@josevalim wasn't the code assuming that the primary key was id before? How would it behave with different keys? got me thinking

Reply:
You are right, the original code assumed that the primary key is id, and I didn't thought about it.

Maybe it makes sense to use the `related_key` of the BelongsTo information?

``` elixir
    case {relation, relation_changed} do
      {%Ecto.Association.BelongsTo{related_key: related_key}, %{} = related} when is_map_key(related, related_key) ->
        Map.put(acc, relation.owner_key, Map.get(related, related_key))
      _ -> acc
    end
```

Reply:
Ah, yes, it was. We should probably add a test then for custom belongs to in this function. In this case, we could rewrite it to:

    case {relation, relation_changed} do
      {%Ecto.Association.BelongsTo{related_key: related_key}, %{^related_key => id}} ->
        Map.put(acc, relation.owner_key, id)
      _ -> acc
    end


You may need to write it as a `with` though.

Reply:
I wrote it using `with`, I tried the match before with the case and didn't work because the variable is not defined before, but I didn't realized to use with.

I also modified the test so it checks that the related_key is being used, if you prefer the test in other way let me know (other belongs_to instead of using the existing Category, for example)